File: ios-app/HealthDecoder/Util/DeviceLanguageDetector.swift
import Foundation
import UIKit

/// Picks the app's initial language from the device's installed keyboards, falling back to the
/// system's preferred locales — the iOS counterpart of `util/DeviceLanguageDetector.kt`.
///
/// Android reads the *currently active* keyboard subtype on the theory that the keyboard a
/// person types in is a better signal of what they comfortably read than the system UI
/// language. iOS exposes no "current keyboard" API to an app that isn't first responder, so
/// this checks the installed keyboards (`UITextInputMode.activeInputModes`) first, then falls
/// back to `Locale.preferredLanguages`, and finally to English.
enum DeviceLanguageDetector {
    private static let isoToSupportedName = [
        "en": "English", "hi": "Hindi", "mr": "Marathi", "gu": "Gujarati",
        "ta": "Tamil", "te": "Telugu", "kn": "Kannada", "bn": "Bengali",
        "pa": "Punjabi", "ml": "Malayalam", "or": "Odia"
    ]

    static func detect() -> String {
        // An installed non-English Indic keyboard is the strongest signal available.
        let keyboardLanguages = UITextInputMode.activeInputModes.compactMap(\.primaryLanguage)
        for tag in keyboardLanguages {
            guard let iso = languageCode(from: tag), let name = isoToSupportedName[iso] else { continue }
            if name != "English" { return name }
        }
        // Then the system's own language ordering.
        for tag in Locale.preferredLanguages {
            if let iso = languageCode(from: tag), let name = isoToSupportedName[iso] { return name }
        }
        // Finally, an English keyboard if that's all that's installed.
        for tag in keyboardLanguages {
            if let iso = languageCode(from: tag), let name = isoToSupportedName[iso] { return name }
        }
        return "English"
    }

    private static func languageCode(from tag: String) -> String? {
        // "hi-IN" / "hi_IN" / "hi" all reduce to "hi".
        tag.split(whereSeparator: { $0 == "-" || $0 == "_" }).first.map(String.init)?.lowercased()
    }
}

File: android-app/app/src/main/java/com/healthdecoder/app/util/DeviceLanguageDetector.kt
package com.healthdecoder.app.util

import android.content.Context
import android.os.Build
import android.view.inputmethod.InputMethodManager
import com.healthdecoder.app.local.AppSettings
import java.util.Locale

/**
 * Picks the app's initial language from the device's currently active keyboard, not the
 * system display language — a user's keyboard is a better signal of the language they're
 * actually comfortable reading (e.g. English system UI with a Hindi keyboard installed).
 * Falls back to English when no keyboard subtype is available or it isn't one of
 * AppSettings.SUPPORTED_LANGUAGES.
 */
object DeviceLanguageDetector {
    private val isoToSupportedName = mapOf(
        "en" to "English",
        "hi" to "Hindi",
        "mr" to "Marathi",
        "gu" to "Gujarati",
        "ta" to "Tamil",
        "te" to "Telugu",
        "kn" to "Kannada",
        "bn" to "Bengali",
        "pa" to "Punjabi",
        "ml" to "Malayalam",
        "or" to "Odia"
    )

    fun detectFromKeyboard(context: Context): String {
        val iso = runCatching {
            val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
            val subtype = imm?.currentInputMethodSubtype
            when {
                subtype == null -> null
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && subtype.languageTag.isNotBlank() ->
                    Locale.forLanguageTag(subtype.languageTag).language
                subtype.locale.isNotBlank() ->
                    Locale(subtype.locale.substringBefore("_")).language
                else -> null
            }
        }.getOrNull()

        return iso?.let { isoToSupportedName[it] } ?: "English"
    }
}
